### PR TITLE
Allow StructuredProfiler to be able to support duplicate columns

### PR DESCRIPTION
**Pull request by ChrisWallace2020**
This required a refactor of `StructuredProfiler._profile` and `report['data_stats']` to both become lists, that are accompanied by mappings (aka schemas) that map column name to a list of indices in the lists corresponding to columns of that name. This refactor also included some test updates, new tests, and documentation changes. This refactor was done in 3 separate PRs into a separate branch for review (#291, #303, #312) that have all been approved and merged, but this PR is to merge all the changes from those into Main, and integrate with new features added since the start of this refactor.

### CODE REVIEW

**Comment by ChrisWallace2020** [comment]
Basically the only new code here pertains to correlations (_get_correlation, _merge_correlation, some tests) as all other code related to the refactor was already approved in a separate branch, but during that time correlation was added to the main branch and was not yet compatible with these new _profile paradigms.

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 1]
```diff
@@ -173,6 +174,8 @@ def test_disabling_columns_during_primitive_diff(self):
         # Test disabled column in both compilers
         compiler2 = col_pro_compilers.ColumnPrimitiveTypeProfileCompiler(data2,
                                                                          options)
+        expected_stddev = (np.sqrt(np.var([-2, -1, 1, 2], ddof=1))
```
Comment (JGSweets) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:177]: add a comment of why

File: dataprofiler/profilers/profile_builder.py [hunk 2]
```diff
@@ -1371,12 +1441,12 @@ def _get_correlation(self, clean_samples):
...
-                if (self._profile[col].profile['data_type'] not in ['int', 'float']
-                        or self._profile[col].null_count > 0):
-                    clean_samples.pop(col)
+            for idx in range(len(self._profile)):
+                if (self._profile[idx].profile['data_type'] not in ['int', 'float']
+                        or self._profile[idx].null_count > 0):
+                    clean_samples.pop(idx)
                 else:
-                    clean_column_ids.append(col_id)
+                    clean_column_ids.append(idx)
```
Comment (ChrisWallace2020) [inline dataprofiler/profilers/profile_builder.py:1449]: This is new

File: dataprofiler/profilers/profile_builder.py [hunk 3]
```diff
@@ -1438,23 +1508,19 @@ def _merge_correlation(self, other):
...
-             for i, profile_name in enumerate(self._profile.keys())
-             if i in col_ids2])
+            [other._profile[idx].profile['statistics']['mean']
+             for idx in range(len(self._profile)) if idx in col_ids2])
         std2 = np.array(
-            [other._profile[profile_name].profile['statistics']['stddev']
-             for i, profile_name in enumerate(self._profile.keys())
-             if i in col_ids2])
+            [other._profile[idx].profile['statistics']['stddev']
+             for idx in range(len(self._profile)) if idx in col_ids2])
```
Comment (ChrisWallace2020) [inline dataprofiler/profilers/profile_builder.py:1523]: This is new

File: dataprofiler/tests/profilers/test_column_profile_compilers.py [hunk 4]
```diff
@@ -186,7 +186,7 @@ def test_disabling_columns_during_primitive_diff(self):
                 'sum': -20.0,
                 'mean': -10.0, 
                 'variance': -46.666666666666664,
-                'stddev': -5.2453259535149215,
+                'stddev': data1.astype(int).std() - data2.astype(int).std(),
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/profilers/test_column_profile_compilers.py:189]: This is new

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 5]
```diff
@@ -308,11 +314,12 @@ def test_correlation(self, *mock):
...
-            [np.nan, 1.0, -0.4907232900448271],
-            [np.nan, -0.4907232900448271, 1.0]
+            [np.nan, 1.0, -0.49072329004483],
+            [np.nan, -0.49072329004483, 1.0]
         ])
-        np.testing.assert_array_equal(expected_corr_mat,
-                                      profiler.correlation_matrix)
+        np.testing.assert_allclose(expected_corr_mat,
+                                   profiler.correlation_matrix,
+                                   rtol=1e-14, atol=1e-14)
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/profilers/test_profile_builder.py:322]: This is new

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 6]
```diff
@@ -342,13 +349,13 @@ def test_merge_correlation(self, *mocks):
...
-            [-0.26559388521279237, 1.0, -0.4907232900448271],
-            [0.26594894270403086, -0.4907232900448271, 1.0]
+            [1.0, -0.265593885212792, 0.265948942704031],
+            [-0.265593885212792, 1.0, -0.49072329004483],
+            [0.265948942704031, -0.49072329004483, 1.0]
         ])
         np.testing.assert_allclose(expected_corr_mat,
                                    merged_profile.correlation_matrix,
-                                   rtol=1e-15, atol=1e-15)
+                                   rtol=1e-14, atol=1e-14)
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/profilers/test_profile_builder.py:358]: This is new

File: dataprofiler/tests/profilers/test_profile_builder.py [hunk 7]
```diff
@@ -359,12 +366,13 @@ def test_merge_correlation(self, *mocks):
...
-            [0.26594894270403086, -0.4907232900448271, 1.0]
+            [1.0, -0.265593885212792, 0.265948942704031],
+            [-0.265593885212792, 1.0, -0.49072329004483],
+            [0.265948942704031, -0.49072329004483, 1.0]
         ])
-        np.testing.assert_array_equal(expected_corr_mat,
-                                      merged_profile.correlation_matrix)
+        np.testing.assert_allclose(expected_corr_mat,
+                                   merged_profile.correlation_matrix,
+                                   rtol=1e-14, atol=1e-14)
```
Comment (ChrisWallace2020) [inline dataprofiler/tests/profilers/test_profile_builder.py:375]: This is new

File: dataprofiler/tests/profilers/test_column_profile_compilers.py (hunk 1 above)
Comment (ChrisWallace2020) [reply dataprofiler/tests/profilers/test_column_profile_compilers.py]: Now its inline using pandas instead.

**Review by JGSweets — APPROVED** [review]

File: dataprofiler/profilers/profile_builder.py [hunk 8]
```diff
@@ -1228,17 +1237,76 @@ def _min_col_samples_used(self):
...
+        :type strict: bool
+        :return: a mapping of indices in schema1 to indices in schema2
+        :rtype: Dict[int, int]
+        """
+
+        len_schema1 = len(schema1)
+        len_schema2 = len(schema2)
+
+        # If both non-empty, must be same length
+        if 0 < len_schema1 != len_schema2 > 0:
```
Comment (gme5078) [inline dataprofiler/profilers/profile_builder.py:1272]: woah. Cool

Comment (ChrisWallace2020) [reply dataprofiler/profilers/profile_builder.py:1272]: Yeah I originally had it as 3 separate `and` conditions and PyCharm told me I could combine them. Btw, this PR has already basically been completely approved except for the chunks that I commented "This is new", so you can focus your review on just those parts. Everything else was already approved into the `duplicate_column_fix` branch. Just letting you know cuz this PR is like 600 lines but only like 60 of them really need review.

**Review by gme5078 — APPROVED** [review]

**Pull request merged**